server/app/controllers/applicant/ApplicantRoutes.java

```diff
@@ -0,0 +1,41 @@
+package controllers.applicant;
+
+import auth.CiviFormProfile;
+import auth.ProfileFactory;
+import play.api.mvc.Call;
+
+/**
+ * Class that computes routes for applicant actions. The route for an applicant may be different
+ * from that for a TI taking action on behalf of an applicant.
+ */
+public final class ApplicantRoutes {
+
+  // There are two cases where we want to use the URL that contains the applicant id:
+  // - TIs performing actions on behalf of applicants.
+  // - The applicant has a profile that does not (yet) include the applicant id.
+  //   This case will eventually go away once existing profiles have expired and been replaced.
+  private static boolean includeApplicantIdInRoute(CiviFormProfile profile) {
+    boolean isTi = profile.isTrustedIntermediary();
+    // TODO(yotommy): add a counter to monitor this case
+    boolean applicantIdInProfile =
+        profile.getProfileData().containsAttribute(ProfileFactory.APPLICANT_ID_ATTRIBUTE_NAME);
+
+    return isTi || !applicantIdInProfile;
+  }
+
+  /**
+   * Returns the URL corresponding to the applicant index action.
+   *
+   * @param profile - Profile corresponding to the logged-in user (applicant or TI).
+   * @param applicantId - ID of applicant for whom the action should be performed.
+   * @return Route for the index action.
+   */
+  public static Call index(CiviFormProfile profile, long applicantId) {
```

Comment:
suggestion (if-minor, non-blocking): Static methods are resistant to extension/refactoring since if there becomes a need for them to depend on something non-static (e.g. the settings manifest) all call sites will need to be updated. Consider making this a public instance method and injecting an instance of `applicantRoutes` where ever it's needed.

Reply:
I had thought about making the instances, but thought it would be a messy refactor. Guice to the rescue! (I had not considered this.) Done, and this simplifies the counter tests.

Reply:
Ah sorry to hassle you on this, but that isn't an improvement for maintainability since with the factory method we can't inject the c'tor. Guice will use the default constructor without you needing to create one, so you can do:

```java
  private final ApplicantRoutes applicantRoutes;

  @Inject
  public ApplicantInformationController(
     ...,
     ApplicantRoutes applicantRoutes
       ) {
    super(profileUtils, versionRepository);
    ...
    this.applicantRoutes = checkNotNull(applicantRoutes); 
  }

  ...
                redirectLink = applicantRoutes.index(profile, applicantId).url();
```

Doing this now just makes it easier to add state in the future. If you'd rather not inject it in all the places that's fine, YAGNI is valid.

Reply:
thanks!